-create repository for Employees

AndreeaMonica opened the pull request:
-create models for Employees
-implement Add/Update/Get methods for Employees
-implement Soft Delete
-add AutoMapper

AndreeaMonica committed:
-create repository for Employees
-create models for Employees
-implement Add/Update/Get methods for Employees
-implement Soft Delete
-add AutoMapper

moisoiu commented on HumanResources/Controllers/EmployeeController.cs:
```diff
@@ -0,0 +1,51 @@
+        [HttpPut("{id}")]
+        public IActionResult UpdateEmployee([FromRoute] long id, [FromBody] UpdateEmployeeRequest updateEmployee)
+        {
+            var employee = employeeRepository.UpdateEmployee(updateEmployee,id);
```
usually is a good practice to add  id at the begin followed by the object

moisoiu commented on HumanResources/Controllers/EmployeeController.cs:
```diff
@@ -0,0 +1,51 @@
+        }
+
+        [HttpDelete("{id}")]
+        public IActionResult SoftDelete([FromRoute] long id)
```
Rename it in DeleteEmployee, no one needs to know how the delete happens behind the scenes

moisoiu commented on HumanResources/Entities/HumanResourceContext.cs:
```diff
@@ -18,14 +14,38 @@ public HumanResourceContext(DbContextOptions<HumanResourceContext> options)
+            return base.SaveChanges();
+        }
+
+        private void UpdateSoftDeleteStatuses() 
```
Try to find a more descriptive name. Because right now the name is confusing with Update and Delete and Statuses. Try to use something maybe like a OverrideDeleteEntity for example.  P.S try to not use that name huehuehue

moisoiu commented on HumanResources/Entities/HumanResourceContext.cs:
```diff
@@ -18,14 +14,38 @@ public HumanResourceContext(DbContextOptions<HumanResourceContext> options)
+                {
+                    case EntityState.Added:
+                        entry.CurrentValues["IsDeleted"] = false;
+                        break;
```
Please check if this case is needed cause right now I have a feeling that this is just extracode that is redundant, cause the default value of bool is false, so theoretically when you add a new Employee for example, the value of IsDeleted will always be False

moisoiu commented on HumanResources/Repositories/EmployeeRepository.cs:
```diff
@@ -0,0 +1,61 @@
+            newEmployee.Name = employee.Name;
+            newEmployee.DateHired = employee.DateHired;
+            newEmployee.LastUpdate = employee.DateHired;
+            newEmployee.IsDeleted = false;
```
This line can be removed, or even better, you can replace it with automapper.

moisoiu commented on HumanResources/Repositories/EmployeeRepository.cs:
```diff
@@ -0,0 +1,61 @@
+            var newEmployee = new Employees();
+            newEmployee.Name = employee.Name;
+            newEmployee.DateHired = employee.DateHired;
+            newEmployee.LastUpdate = employee.DateHired;
```
I assume that the LastUpdate shouldn't be the DateHired, should be the current time the record was inserted.  

P.S I think you can do something on the override **SaveChanges** with this one :D

moisoiu commented on HumanResources/Repositories/EmployeeRepository.cs:
```diff
@@ -0,0 +1,61 @@
+            return getEmployee;
+        }
+
+        public bool SoftDelete(long id)
```
Rename the method with a more descriptive name

moisoiu commented on HumanResources/Repositories/EmployeeRepository.cs:
```diff
@@ -0,0 +1,61 @@
+            var employee = context.Employees.FirstOrDefault(x => x.Id == id);
+            context.Remove(employee);
+            context.SaveChanges();
+            return true;
```
How do you know the Delete was true, what happens if you dont find an employee? you will return a default true no matter what

moisoiu commented on HumanResources/Repositories/EmployeeRepository.cs:
```diff
@@ -0,0 +1,61 @@
+        public long UpdateEmployee(UpdateEmployeeRequest updateEmployee, long id)
+        {
+            var employee = context.Employees.FirstOrDefault(x => x.Id == id);
+            employee.LastUpdate = updateEmployee.LastUpdate;
```
LastUpdate shouldn't be decided by who is using the application, because people will might override the time. It should be treated by the server. Server should do the LastUpdate time

AndreeaMonica committed:
Applied changes :
-renamed method DeleteEmployee
-renamed method OverrideDelete
-removed redundant code
-used AutoMapper for AddEmployee
-added automatically DateTime for LastUpdate and FirstUpdate

moisoiu commented on HumanResources/Entities/HumanResourceContext.cs:
```diff
@@ -18,14 +14,38 @@ public HumanResourceContext(DbContextOptions<HumanResourceContext> options)
             {
+                base.OnModelCreating(modelBuilder);
+                modelBuilder.Entity<Employees>()
+                            .HasQueryFilter(m => EF.Property<bool>(m,"IsDeleted")==false);
```
You can use !EF.Property<bool>(m,"IsDeleted") instead of EF.Property<bool>(m,"IsDeleted")==false

moisoiu commented on HumanResources/Repositories/EmployeeRepository.cs:
```diff
@@ -21,13 +21,9 @@ public EmployeeRepository(HumanResourceContext context, IMapper mapper)
-            //var newEmployee = mapper.Map<Employees>(employee);
+            var newEmployee = mapper.Map<Employees>(employee);
+            newEmployee.DateHired = DateTime.Now;
+            newEmployee.LastUpdate = newEmployee.DateHired;
```
I think LastUpdate should be DateTime now and not DateHired

AndreeaMonica committed:
Applied changes for LastUpdate and  method
